homeassistant/components/withings/common.py

```diff
@@ -89,184 +370,278 @@ def __init__(
+
+    persistent_notification.async_dismiss(
+        hass=hass, notification_id=const.WEBHOOK_NOTIFICATION_ID
+    )
+
+    return WebhookAvailability.SUCCESS
+
+
 class WithingsDataManager:
     """A class representing an Withings cloud service connection."""
 
-    service_available = None
-
-    def __init__(self, hass: HomeAssistant, profile: str, api: ConfigEntryWithingsApi):
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        profile: str,
+        api: ConfigEntryWithingsApi,
+        user_id: int,
+        webhook_config: WebhookConfig,
+    ):
         """Initialize data manager."""
         self._hass = hass
         self._api = api
+        self._user_id = user_id
         self._profile = profile
-        self._slug = slugify(profile)
+        self._webhook_config = webhook_config
 
-        self._measures = None
-        self._sleep = None
-        self._sleep_summary = None
+        self._is_available = True
+        self._cancel_interval_update_interval: Optional[CALLBACK_TYPE] = None
+        self._cancel_configure_notify_interval: Optional[CALLBACK_TYPE] = None
+        self._api_notification_id = f"withings_{self._user_id}"
 
-        self.sleep_summary_last_update_parameter = None
-        self.throttle_data = {}
+        self._update_all_delay = datetime.timedelta(seconds=randint(0, 20))
+        self._update_all_interval = datetime.timedelta(minutes=30)
+        self._subscribe_webhoook_delay = datetime.timedelta(seconds=randint(30, 60))
+        self._subscribe_webhoook_interval = datetime.timedelta(minutes=120)
 
-    @property
-    def profile(self) -> str:
-        """Get the profile."""
-        return self._profile
-
-    @property
-    def slug(self) -> str:
-        """Get the slugified profile the data is for."""
-        return self._slug
+        self._notify_subscribe_delay = datetime.timedelta(seconds=5)
 
     @property
-    def api(self) -> ConfigEntryWithingsApi:
-        """Get the api object."""
-        return self._api
+    def webhook_config(self) -> WebhookConfig:
+        """Get the webhook config."""
+        return self._webhook_config
+
+    def _do_interval(
+        self, action: Callable[[], Coroutine], delay: timedelta, interval: timedelta
+    ) -> Callable:
+        cancel_func1 = None
+        cancel_func2 = None
+
+        @callback
+        def runner(now: datetime.datetime) -> None:
+            self._hass.async_run_job(action())
+
+        def cancel() -> None:
+            nonlocal cancel_func1
+            nonlocal cancel_func2
+            cancel_func1()
+            cancel_func2()
+
+        cancel_func1 = async_track_point_in_utc_time(
+            self._hass, action=runner, point_in_time=dt.utcnow() + delay,
+        )
 
-    @property
-    def measures(self) -> MeasureGetMeasResponse:
-        """Get the current measures data."""
-        return self._measures
+        cancel_func2 = async_track_time_interval(self._hass, runner, interval + delay)
```

Comment:
Won't this fire it twice at delay? What is wrong with just using `async_track_time_interval`? I think this code could use some docstrings and comments :)

Reply:
Added more docs to help explain what's going on. This works like `async_track_time_interval` except this will run the function immediately and then set the interval. This is beneficial as the interval for data updates is 30 min and webhook reconfig 2hr.
The other benefit we get from this is adding a random delay before starting. This help avoid too many requests from hitting the withings API at once from this HA instance and across the globe. I've noticed the withings API can be a bit finicky if hit too hard.